### ISSUE TITLE
No error on duplicate alias

### ISSUE THREAD

**Issue by mcorner** [BranchMetrics/react-native-branch-deep-linking#265]
I am attempting to provide vanity urls for sharesheets using the alias feature.  I would like to ensure that two users don't try to share with the same alias.

#84 implies that I would get a 409 if two users open the sharesheet with the same alias, but this doesn't seem to happen.

This is how I am creating the link in the app:

```
    let branchUniversalObject = await Branch.createBranchUniversalObject('canonicalIdentifier', {
      automaticallyListOnSpotlight: true,
      metadata: { },
      title: 'User share',
      contentDescription: 'User share in app'});

    let shareOptions = { messageHeader: 'Foobar', messageBody: 'Foobar' };
    let linkProperties = { feature: 'share', channel: 'app', alias: 'mark' };
    let controlParams = { };
    let {channel, completed, error} = await branchUniversalObject.showShareSheet(shareOptions, linkProperties, controlParams);

    console.log("SHOW SHARE: " + error);
```

But if two different users (both are logged in, but with different email addresses and set with Banch.setIdentity) pop the share sheet with this same alias "mark", then I don't get an error.  Am I doing something wrong?

**Comment by jdee** [comment]
The honest answer is I have no idea. I can see what I wrote in January, when I was new to the project, and I remember adding the missing params. As for what the behavior of the Branch system is at any given time, what I have to say about it in this repo should not be taken too seriously. The support team would be much better able to handle this sort of issue. If you open a support ticket by emailing integrations@branch.io or via https://support.branch.io, there's a team of people who know a lot of this stuff far better than I do. To answer questions like this, I usually go to them, and my involvement takes time away from maintaining the SDK.

I can tell you a few things: Any error that is passed back from the native clients in calling `generateShortUrl` or `showShareSheet` is returned to JS. The issue in #84 used `generateShortUrl` rather than `showShareSheet`. The behavior of those methods in the native SDKs is a bit different. For example, `showShareSheet` does not return the generated URL to the client. And of course you can always edit any link after it's been created, but I don't know if that is accomplished by calling `generateShortUrl` or `showShareSheet` more than once or what support the client libraries offer.

**Comment by mcorner** [comment]
Thanks, I contacted support.  The basic answer is that you can't use aliases with the referral system.  Nor should you really have users creating aliased links to share.  Duplicates can't or are hard to prevent.

Our current plan is to create our own vanity URLs to forward to the referral links created in the app.  Hopefully that will work.

**Comment by jdee** [comment]
Glad you got it sorted out, @mcorner. I knew support would know the answer.

**Comment by mcorner** [comment]
@jdee After a bunch of back and forth with support about alias links and referrals, they have confirmed that there is some kind of bug in the sdk.  They said they would get in contact with the developers (I guess you :)

I was wrong about them saying that you can't use aliases with the referral system.  You can.

But if you use an alias in the sdk then the resulting branch link does not contain a branch identity id, which prevents the referral credit system from working.  If you create the same link in curl, it does work.  For now we are creating links using direct API calls (we do it server side now) as a work around.

Thanks, and I imagine a new issue will get opened up.

-Mark

**Comment by jdee** [comment]
@mcorner The issue with the `$identity_id` is in the iOS SDK (it works fine on Android). The RN SDK doesn't really play any role at that level of code. It just wraps the others. The issue is being addressed and will be fixed in an upcoming release of the iOS SDK. Once that's done, the new iOS SDK will be included here and released.

Regarding the original issue with the error code, there are also some issues with returning that on iOS. The behavior of `showShareSheet` on iOS is to share a long URL if there is any error from the API in creating a short URL, including the case when the alias is taken.

Anyway, if your purpose is to let users claim aliases as vanity URLs and then prompt them if the alias is already taken, it's probably best to use `generateShortUrl` rather than `showShareSheet`. The idea would be something like:

```js
import { Share } from 'react-native'

try {
  let { url } = await this.buo.generateShortUrl(linkProperties, controlParams)

  // OK. Allocate this URL to the user and continue
  Share.share({
    title: 'My vanity link',
    message: url
  })
}
catch (error) {
  if (error.code == 'RNBranch::Error::DuplicateResourceError') {
    // prompt the user for a different alias and try again
  }
}
```

This gives you greater flexibility in handling duplicates and deciding how to handle it in the UI when there is a conflict.

**Comment by mcorner** [comment]
We switched to generating the links on the server side (direct call to the public API) and returning those to the client.  Then we use the builtin sharesheet from react-native.  This is working well for us now.  (One suggestion: allow alias links to be changed, the immutability is painful when you think about changing the open graph metadata etc)

**Comment by jdee** [comment]
@mcorner The issue with missing identity IDs in aliases has been fixed with iOS SDK 0.19.5, which is now available in the master branch. This will be released soon. There's another Android release that would be nice to squeeze in if it's available soon enough.

I think creating Branch links on the server makes a lot of sense. If that's working for you, then I guess this isn't so urgent for you, but we'll get it released soon.

If you haven't sent your suggestion to support already, that would be a good thing to share with them. Maybe there's a way, and they can tell you the secret. One way or another, I'm sure they can help you improve whatever's not quite right.
